File: LC/Python/Medium/2130.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        ans = 0
        slow, fast = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            
        curr, prev = slow, None
        while curr:
            curr.next, prev, curr = prev, curr, curr.next
        
        start = head
        while prev:
            ans = max(ans, start.val + prev.val)
            start = start.next
            prev = prev.next
            
        return ans

File: LC/C++/Medium/2130.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int pairSum(ListNode* head) {
        ListNode* s = head;
        ListNode* f = head;

        while (f && f->next) {
            s = s->next;
            f = f->next->next;
        }
        ListNode* prev = nullptr;
        while (s) {
            ListNode* next = s->next;
            s->next = prev;
            prev = s;
            s = next;
        }

        int ans = 0;
        ListNode* p1 = head;
        ListNode* p2 = prev;
        while (p2) {
            ans = std::max(ans, p1->val + p2->val);
            p1 = p1->next;
            p2 = p2->next;
        }

        return ans;
    }
};